src/urljoin.php

```diff
@@ -18,13 +18,34 @@ function urljoin($base, $rel) {
 		return $base;
 	}
 
+	$is_valid_scheme_fn = function ($scheme) {
+		$valid_scheme_chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789+-.';
+
+		for($i = 0; $i<strlen($scheme); $i++) {
+			if (strpos($valid_scheme_chars, $scheme[$i]) === false) {
+				return false;
+			}
+		}
+
+		return true;
+	};
+
 	$uses_relative = array('', 'ftp', 'http', 'gopher', 'nntp', 'imap',
 		'wais', 'file', 'https', 'shttp', 'mms',
 		'prospero', 'rtsp', 'rtspu', 'sftp',
 		'svn', 'svn+ssh', 'ws', 'wss');
 
+	$scheme = '';
+	if (strpos($rel, ':') !== false) {
+		$scheme = explode(':', $rel)[0];
+	}
+
 	$pbase = parse_url($base);
-	$prel = parse_url($rel);
+	if ($scheme && false === $is_valid_scheme_fn($scheme)) {
```

Comment:
Per the docs on `parse_url`, if the scheme is malformed it returns false. So, what would have been better is this:

```suggestion
    $prel = parse_url($rel);
    if ($prel === false) {
        $prel = array('path' => $rel);
    }
```

At least on my quick smoke tests that appears to work on the test cases you added to the JSON file.

Reply:
Okay this new 'suggested change' feature seems to make mincemeat of the formatting.

In any case the root issue is just that `parse_url` is doing the right thing and returning `false` for these things that look like malformed URL schemes, and so we just need to detect that `$prel===false` and create the minimal `path` array instead in that case.